### PR TITLE
feat: add list nodes & create project

### PR DESCRIPTION
**Pull request by kai0x**

### CODE REVIEW

File: cli.js [hunk 1]
```diff
@@ -82,6 +87,104 @@ yargs(hideBin(process.argv))
...
+            let url = defaultNetworks[argv.network].url
+            const provider = new ethers.providers.JsonRpcProvider(url);
+            const contract = new ethers.Contract(ArmadaNodesABI.address, ArmadaNodesABI.abi, provider);
+
+            const nodes = await contract.getNodes(argv.operatorIdOrZero, argv.topology, argv.skip, argv.size);
+            console.log(nodes);
+        }
+    )
+    .command(
+        'project create <network> <name> <email> <content> <checksum> <wallet-password> <wallet-address>',
```
Comment (kai0x) [inline cli.js:132]: these seem like a lot of arguments. i wonder if there is a way to make cleaner or maybe provide a few examples with flags: https://oclif.io/docs/flags

File: abis/ArmadaNodes.json [hunk 2]
```diff
@@ -0,0 +1,1547 @@
+{
```
Comment (sbaranov) [inline abis/ArmadaNodes.json:1]: FYI Eventually this should use git submodule to grab abis. This doesn't have to happen in this PR though.

**Comment by kai0x** [comment]
actually going to convert this back to a draft for now to move the keystore to `/.armada/keystore` location

File: networks.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
+
+// For adding when TS is setup
+// export interface NetworkInfo {
+//     url: string
+//     id?: number
+//     label?: string
+//   }
+// export type Networks = { [name: string]: NetworkInfo }
+
+export const defaultNetworks = {
```
Comment (sbaranov) [inline networks.js:10]: FYI we currently have staging and testnet. They're both goerli right now.

File: cli.js [hunk 4]
```diff
@@ -82,6 +87,104 @@ yargs(hideBin(process.argv))
             console.log(`Keystore generated for: ${wallet.address}`);
         }
     )
+    .command(
+        'nodes list',
+        'List all nodes <network> <operator-id-or-zero> <topology> <skip> <size>',
```
Comment (sbaranov) [inline cli.js:92]: FYI The network should be set to testnet by default. It should be possible to switch to staging by using --staging.

File: cli.js (hunk 1 above)
Comment (sbaranov) [inline cli.js:132]: FYI Eventually, this should save the password in keychain so that we doesn't prompt for it on every run. 

https://www.npmjs.com/package/keychain
https://stackoverflow.com/questions/60985877/how-can-i-store-a-program-password-in-the-os-keystore

File: abis/ArmadaNodes.json (hunk 2 above)
Comment (kai0x) [reply abis/ArmadaNodes.json:1]: sounds good! good to know we'll automate this

File: networks.js (hunk 3 above)
Comment (kai0x) [reply networks.js:10]: sounds good, I'll add a key for staging and testnet 👍

File: cli.js (hunk 1 above)
Comment (kai0x) [reply cli.js:138]: interesting so I dug around to see if i could find an npm that stores to keychain in all systems window/linux/mac. found this: https://www.npmjs.com/package/keyring not sure how good it is. i'll test it out 👍

**Comment by sbaranov** [comment]
I looked in the code, it's not using the OS keychain, so it's not secure.
The OS will prevent other apps from accessing the armada passwords.

File: cli.js (hunk 4 above)
Comment (kai0x) [reply cli.js]: sounds good, just added testnet as the default here: https://github.com/dispersed-labs/dispersed-cli/pull/12/files#diff-180b3d2af89416b99563fff8c37eb8ca1ef7591fc1ddd736b5dacf740ccd3760R107

File: cli.js (hunk 1 above)
Comment (kai0x) [reply cli.js:155]: actually i looked into https://www.npmjs.com/package/keyring and not sure on that one since the usage is quite low.

this other one looks like it supports mac keychain and windows/linux: https://www.npmjs.com/package/keytar and has high usage
 
i'll test adding in a keychain setup as a new pr after these two prs are merged 👍

File: networks.js [hunk 5]
```diff
@@ -13,16 +13,6 @@ export const defaultNetworks = {
...
-      url: 'https://ropsten.infura.io/v3/76fb6c10f1584483a45a0a28e91b07ad',
-      id: 3,
-      label: 'Ropsten',
-    },
-    rinkeby: {
-      url: 'https://rinkeby.infura.io/v3/76fb6c10f1584483a45a0a28e91b07ad',
-      id: 4,
-      label: 'Rinkeby',
-    },
     goerli: {
```
Comment (sbaranov) [inline networks.js:16]: Mainnet and goerli can be removed as we don't have such deployments.

File: cli.js (hunk 1 above)
Comment (kai0x) [reply cli.js:155]: oh just saw your comment below! yep sounds good it doesn't look like keyring is sufficient

File: networks.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    goerli: {
+      url: 'https://goerli.infura.io/v3/76fb6c10f1584483a45a0a28e91b07ad',
+      id: 5,
+      label: 'Görli',
+    },
+    // testnet and staging both use goerli
+    testnet: {
+      url: 'https://goerli.infura.io/v3/76fb6c10f1584483a45a0a28e91b07ad',
+      id: 5,
+      label: 'Görli',
```
Comment (sbaranov) [inline networks.js]: The label doesn't seem to be used for anything for anything. Remove?

File: cli.js [hunk 7]
```diff
@@ -90,6 +96,131 @@ yargs(hideBin(process.argv))
             console.log(`Keystore generated for: ${wallet.address}`);
         }
     )
+    .command(
+        'nodes list <network> <operator-id-or-zero> <skip> <size> <topology>',
+        'List all nodes',
+        (yargs) => {
+            return yargs
+                .option('network', {
```
Comment (sbaranov) [inline cli.js:104]: The network should default to testnet, unless --staging is specified. No need for positional parameter :-)

File: cli.js (hunk 1 above)
Comment (sbaranov) [reply cli.js:155]: Sounds good!

File: cli.js [hunk 8]
```diff
@@ -90,6 +96,131 @@ yargs(hideBin(process.argv))
...
+                    describe: 'The number of nodes to skip. Example: 0',
+                    default: 0,
+                    type: 'number',
+                })
+                .option('size', {
+                    describe: 'The number of nodes to return. Example: 100',
+                    default: 100,
+                    type: 'number',
+                })
+                .option('topology', {
```
Comment (sbaranov) [inline cli.js:123]: Let's make this a flag since it should just be false by default (unless --topology is specified).

File: cli.js [hunk 9]
```diff
@@ -90,6 +96,131 @@ yargs(hideBin(process.argv))
...
+        'nodes list <network> <operator-id-or-zero> <skip> <size> <topology>',
+        'List all nodes',
+        (yargs) => {
+            return yargs
+                .option('network', {
+                    describe: 'The network to use. Example: mainnet, testnet, staging, goerli, rinkeby',
+                    type: 'string',
+                    default: 'testnet',
+                })
+                .option('operator-id-or-zero', {
```
Comment (sbaranov) [inline cli.js:109]: I think this should be a flag instead (--operator 0x123) and should default to zero.

File: cli.js [hunk 10]
```diff
@@ -90,6 +96,131 @@ yargs(hideBin(process.argv))
...
+        async (argv) => {
+            console.log(`Listing all nodes.`);
+
+            let url = defaultNetworks[argv.network].url
+            const provider = new ethers.providers.JsonRpcProvider(url);
+
+            let address;
+            let abi;
+
+            if (argv.network === 'mainnet') {
```
Comment (sbaranov) [inline cli.js:138]: Looks like this branch can be removed now.

File: cli.js [hunk 11]
```diff
@@ -90,6 +96,131 @@ yargs(hideBin(process.argv))
...
+                    describe: 'The address of the wallet',
+                    type: 'string',
+                })
+        },
+        async (argv) => {
+            console.log(`Creating project: ${argv.name}`);
+
+            let address;
+            let abi;
+            if (argv.network === 'mainnet') {
```
Comment (sbaranov) [inline cli.js:193]: Remove this branch

**Comment by kai0x** [comment]
cleaned up these items to use flags :)

File: README.md [hunk 12]
```diff
@@ -15,6 +15,25 @@ npm install dispersed-cli
...
+$ npx dispersed nodes list 0 100 false
+
+# lists top 100 topology nodes in testnet
+$ npx dispersed nodes list 0 100 false --topology
+
+# lists top 100 non-toplogy nodes in staging
+$ npx dispersed nodes list 0 100 false --staging
+
+# filter by specific operator
+$ npx dispersed nodes list 0 100 false --staging --operator 0x0000000000000000000000000000000000000000000000000000000000000123
```
[the remark this answers is not in the document]
Comment (sbaranov) [reply README.md:29]: I'm confused, what does the "false" correspond to? The syntax should be `npx dispersed nodes list [skip [size]] [--topology]`. So, skip and size should be optional.

File: README.md [hunk 13]
```diff
@@ -15,6 +15,25 @@ npm install dispersed-cli
...
+$ npx dispersed nodes list 0 100 false --operator 0x0000000000000000000000000000000000000000000000000000000000000123
+
+# create project in testnet
+$ npx dispersed project create kai kai@gmail.com https://url-to-your-frontend-files.com 0x0000000000000000000000000000000000000000000000000000000000000000 somepassword 0xb77c81c43C461528FFeA65EbDf5a17691247185A
```
Comment (sbaranov) [inline README.md:33]: Only name/email should be required, and content/checksum should be optional. The address should probably default to the first key in keystore (should be able to override with a flag --address). And password should be prompted. It's OK to do some of these in another PR if you want.

File: README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (kai0x) [reply README.md]: ah the false was from when the topology value wasn't optional

odd that commands still works even with false in there (it just doesn't use it). i went ahead and removed it from the examples

File: README.md (hunk 13 above)
Comment (kai0x) [reply README.md:33]: @sbaranov sounds good 

a) I updated the wallet/address to be prompted. Here's what it looks like: https://www.loom.com/share/cc07eda803d54d079915e864cdb116f2 

the user can pick and choose which wallet they want to use, otherwise the first choice is just the first keystore in the folder

b) for the optional content/checksum what would the default be if those are optional? the `createProject` solidity function requires these two right?

Comment (sbaranov) [reply README.md:33]: The contract's createProject allows empty bytes[] for content/checksum.

Comment (kai0x) [reply README.md:33]: ah sounds good, wasn't sure if it was an error that these aren't needed earlier. went ahead and add a default value of empty bytes for checksum and empty string for content

File: README.md [hunk 14]
```diff
@@ -15,6 +15,25 @@ npm install dispersed-cli
 ```bash
 # saves wallet for signing transactions
 $ npx dispersed key import
+
+# lists top 100 non-topology armada nodes in testnet
+$ npx dispersed nodes list 0 100
```
Comment (sbaranov) [inline README.md:20]: Instead of 0 and 100, it's --skip 0 --size 100, right?

File: README.md (hunk 13 above)
Comment (sbaranov) [reply README.md:33]: > a) I updated the wallet/address to be prompted. Here's what it looks like:

Looks nice!

**Review by sbaranov — APPROVED** [review]

File: README.md (hunk 14 above)
Comment (kai0x) [reply README.md:20]: ah so for `positional` required inputs it doesn't have flags. let me see if there is a workaround

Comment (sbaranov) [reply README.md:20]: Instead of positional required they can just be optional flags

Comment (kai0x) [reply README.md:20]: sg 👍 I'll move it to optional
